internal/config/translate.go

```diff
@@ -15,360 +15,29 @@
 package config
 
 import (
+	"github.com/coreos/ignition/config/translate"
 	from "github.com/coreos/ignition/config/v3_0_experimental/types"
 	"github.com/coreos/ignition/internal/config/types"
 )
 
-func intToPtr(x int) *int {
-	return &x
-}
+var (
+	tr translate.Translator
+)
 
-func strToPtr(s string) *string {
-	if s == "" {
-		return nil
-	}
-	return &s
-}
+func translateIgnition(old from.Ignition) types.Ignition {
+	ret := types.Ignition{}
+	tr.Translate(&old.Config, &ret.Config)
+	tr.Translate(&old.Timeouts, &ret.Timeouts)
+	tr.Translate(&old.Security, &ret.Security)
```

Comment:
Hmm, OK I'm not sure I understand how this works. Why do we need a custom translator for the `Ignition` bit only? Is it because the rest of the spec is only ever a superset of previous versions so the default translation behaviour just works?

Reply:
We need a custom translator for the Ignition bit only since it's the only bit that has a field that changed. I couldn't think of a good way to allow specifying custom translators on the field level.

We call the translator on the other parts of the `Ignition` struct since those bits are the same.

Worth noting that the version in this case is actually the same, but the test code accidentally omitted the version from some of the test cases which ended up working out because we can test that it writes the new version correctly (since it would fail if it copied the empty string).